mesos-rxjava-core/src/main/java/org/apache/mesos/rx/java/MesosSchedulerClientBuilder.java

```diff
@@ -0,0 +1,164 @@
+     * @return this builder (allowing for further chained calls)
+     */
+    @NotNull
+    public MesosSchedulerClientBuilder<Send, Receive> receiveCodec(
+        @NotNull final MessageCodec<Receive> receiveCodec
+    ) {
+        this.receiveCodec = receiveCodec;
+        return this;
+    }
+
+    /**
+     * @param subscribe     The {@link org.apache.mesos.v1.scheduler.Protos.Call.Type#SUBSCRIBE}
+     *                      {@link org.apache.mesos.v1.scheduler.Protos.Call} to be sent to Mesos
+     *                      when opening the event stream.
+     * @return this builder (allowing for further chained calls)
+     */
+    @NotNull
+    public MesosSchedulerClientBuilder<Send, Receive> subscribe(
+        @NotNull final Send subscribe
+    ) {
+        this.subscribe = subscribe;
+        return this;
+    }
+
+    /**
+     * This method provides the means for a user to define how the event stream will be processed.
+     * <p/>
+     * The function passed to this method will function as the actual event processing code for the user.
+     * <p/>
+     * The stream the users will source from is an {@link Observable} of {@code Receive}s. With this stream
+     * source a number of functions can be applied to transform/interact/evaluate the stream.
+     * <p/>
+     * The output of this function represents the users reaction to each event represented as an
+     * {@code Observable<Optional<SinkOperation<Send>>>}. If {@link Optional#isPresent()} the specified
+     * {@link SinkOperation} will be processed.
+     * <p/>
+     * For example, if you wanted to log all tasks that result in error:
+     * <pre><code>
+     * events -> {
+     *     final Observable&lt;Optional&lt;SinkOperation&lt;Call>>> errorLogger = events
+     *         .filter(event -> event.getType() == Event.Type.UPDATE && event.getUpdate().getStatus().getState() == TaskState.TASK_ERROR)
+     *         .doOnNext(e -> LOGGER.warn("Task Error: {}", ProtoUtils.protoToString(e)))
+     *         .map(e -> Optional.empty());
+     * }
+     * </code></pre>
+     * @param streamProcessing    The function that will be woven between the event spout and the call sink
+     * @return this builder (allowing for further chained calls)
+     */
+    @NotNull
+    public MesosSchedulerClientBuilder<Send, Receive> processStream(
+        @NotNull final Function<Observable<Receive>, Observable<Optional<SinkOperation<Send>>>> streamProcessing
+    ) {
+        this.streamProcessor = streamProcessing;
+        return this;
+    }
+
+    /**
+     * Builds the instance of {@link MesosSchedulerClient} that has been configured by this builder.
+     * All items are expected to have non-null values, if any item is null and exception will be thrown.
+     * @return The configured {@link MesosSchedulerClient}
+     */
+    @NotNull
+    public final MesosSchedulerClient<Send, Receive> build() {
+        return new MesosSchedulerClient<>(
+            mesosUri,
```

Comment:
Good catch, at one point I had that sent in the constructor and enforced it there.